### PR TITLE
Fix some template definitions for clang 12

### LINKED ISSUE #2370: [macOS] "Symbol not found" in release builds
As discovered by @mweisgut, hyriseTest fails for release builds on current Mac systems (clang12, macOS 11.4).

```
[----------] 38 tests from ExpressionEvaluatorToValuesTest
[ RUN      ] ExpressionEvaluatorToValuesTest.TernaryOrLiterals
dyld: lazy symbol binding failed: Symbol not found: __ZN7opossum19ExpressionEvaluator29evaluate_expression_to_resultIiEENSt3__110shared_ptrINS_16ExpressionResultIT_EEEERKNS_18AbstractExpressionE
  Referenced from: /Users/julianmenzler/workspace_hyrise/repo_hyrise2/cmake_release_deb/./hyriseTest
  Expected in: flat namespace

dyld: Symbol not found: __ZN7opossum19ExpressionEvaluator29evaluate_expression_to_resultIiEENSt3__110shared_ptrINS_16ExpressionResultIT_EEEERKNS_18AbstractExpressionE
  Referenced from: /Users/julianmenzler/workspace_hyrise/repo_hyrise2/cmake_release_deb/./hyriseTest
  Expected in: flat namespace

Abort trap: 6
```

### PR DESCRIPTION
**Pull request by Bensk1**
Fixes #2370
 
Apparently, clang12 (default when installing llvm via brew on macOS) causes some trouble with template definitions for release builds. This PR should fix the issue. It also removes the unnecessary template of `JoinHash::calculate_radix_bits()`.

(Compile time difference on nemea was negligible.)

### CODE REVIEW

**Comment by Bouncner** [comment]
I am working (for quite a while already to be honest) on the radix stuff and the template argument makes sense to estimate the hash map size. I am fine with removing it for now, but I am curious: is there any other reason to remove than "it' just not needed"?

**Comment by Bensk1** [comment]
> I am working (for quite a while already to be honest) on the radix stuff and the template argument makes sense to estimate the hash map size. I am fine with removing it for now, but I am curious: is there any other reason to remove than "it' just not needed"?

No, not really. I had to move its definition from .cpp to .hpp in #2360 because gcc10 complained. When discussion the clang12 template issues with Markus, he saw that `calculate_radix_bits` does never use the template. Because of that, I removed the template and moved the implementation back to the .cpp file. Do I understand correctly that you are planning to use the template parameter?

**Comment by Bouncner** [comment]
Yes, I use it in a branch that is blocked by some other branch. I don't know if I will ever open a PR for them. So no reason to not remove the template now. I was just curious if there was a more "technical reason".

**Review by Bouncner — APPROVED** [review]

**Comment by mweisgut** [comment]
I cannot run `hyriseTest` (`RelWithDebInfo` build) successfully (using brew llvm 12):
```
[ RUN      ] ExpressionEvaluatorToValuesTest.TernaryOrLiterals
dyld: lazy symbol binding failed: Symbol not found: __ZN7opossum19ExpressionEvaluator29evaluate_expression_to_resultIiEENSt3__110shared_ptrINS_16ExpressionResultIT_EEEERKNS_18AbstractExpressionE
  Referenced from: <path>/bRelWithDebInfo/./hyriseTest
  Expected in: flat namespace

dyld: Symbol not found: __ZN7opossum19ExpressionEvaluator29evaluate_expression_to_resultIiEENSt3__110shared_ptrINS_16ExpressionResultIT_EEEERKNS_18AbstractExpressionE
  Referenced from: <path>/bRelWithDebInfo/./hyriseTest
  Expected in: flat namespace

[1]    15931 abort      ./hyriseTest
```

@Bensk1 I'm not sure which commit I successfully build and run in the afternoon. Based on the timestamps, I assume it was 6028d66 or 453cd70.

**Comment by Bensk1** [comment]
> Edit: [6028d66](https://github.com/hyrise/hyrise/commit/6028d669c8e60829d19c63dbeea1d36f044c3385) doesn't result in the above error and runs successfully, [453cd70](https://github.com/hyrise/hyrise/commit/453cd703f97184234fdcd36950aa27b779877123) does result in the above error.

This is really strange because 6028d66 does not change anything except removing two comments.

**Comment by mweisgut** [comment]
> > Edit: [6028d66](https://github.com/hyrise/hyrise/commit/6028d669c8e60829d19c63dbeea1d36f044c3385) doesn't result in the above error and runs successfully, [453cd70](https://github.com/hyrise/hyrise/commit/453cd703f97184234fdcd36950aa27b779877123) does result in the above error.
> 
> This is really strange because [6028d66](https://github.com/hyrise/hyrise/commit/6028d669c8e60829d19c63dbeea1d36f044c3385) does not change anything except removing two comments.

But 6028d66 contains the `ExpressionEvaluator` template definitions whereas 453cd70 does not. So with the template definitions, the above error doesn't occur, without the template definitions, the error occurs.

I will try to build it with llvm 12.0.0_1 and get back to you.

**Comment by Bensk1** [comment]
Ah, I misread the commit history on GitHub. It shows newest commits at the bottom, the opposite way of `git log`.

**Comment by mweisgut** [comment]
I also tried (multiple times) to build the latest version of this branch using llvm 12.0.0_1 and the `dyld: lazy symbol binding failed` error in `ExpressionEvaluatorToValuesTest.TernaryOrLiterals` still occurred. Using version 6028d66, `hyriseTest` runs without any errors.

**Comment by mweisgut** [comment]
#2370

**Comment by Bensk1** [comment]
Should work now. @mweisgut, could you give it another try?

**Comment by mweisgut** [comment]
`hyriseTest` successfully runs now in `Release` and `RelWithDebInfo` modes.

**Review by mweisgut — APPROVED** [review]

File: src/lib/expression/evaluation/expression_evaluator.cpp [hunk 1]
```diff
@@ -1613,4 +1613,14 @@ std::vector<std::shared_ptr<ExpressionResult<Result>>> ExpressionEvaluator::_pru
   return results;
 }
 
+// We explicitly instantiate these template functions because (at least) clang-12 does not instantiate them for us.
```
Comment (julianmenzler) [inline src/lib/expression/evaluation/expression_evaluator.cpp:1616]: Maybe also mention the platform macOS here?

Comment (mweisgut) [reply src/lib/expression/evaluation/expression_evaluator.cpp:1616]: For this, we should verify that the error does not appear on ubuntu with clang 12.

Comment (julianmenzler) [reply src/lib/expression/evaluation/expression_evaluator.cpp:1616]: Does the CI use clang 12?

Comment (Bensk1) [reply src/lib/expression/evaluation/expression_evaluator.cpp:1616]: > Does the CI use clang 12?

No, it doesn't.

Comment (julianmenzler) [reply src/lib/expression/evaluation/expression_evaluator.cpp:1616]: Good. Let's leave the comment as is and merge this PR.

Comment (Bouncner) [reply src/lib/expression/evaluation/expression_evaluator.cpp:1616]: Any reason not to keep clang9 and use clang12 over clang10 in the CI?

Comment (Bensk1) [reply src/lib/expression/evaluation/expression_evaluator.cpp:1616]: > Any reason not to keep clang9 and use clang12 over clang10 in the CI?

Not really but we might have to upgrade the Ubuntu version then. Also, with the Mac ARM CI we will have clang12 as part of the CI.

**Pull request merged**